### PR TITLE
include dark current contribution in M2 pedestal constraint for Phase2

### PR DESCRIPTION
**Pull request by kpedro88**
When SiPM aging is enabled, the dark current (an extra source of electronics noise, separate from the regular pedestal) can increase drastically. The pedestal constraint in HCAL Method 2 reconstruction did not account for this increase, leading to biased energy for some RecHits. Now it is included, but only enabled for Phase2. An additional parameter is added to the Method 2 parameter set, which must eventually be propagated to HLT separately (since M2 parameters are not kept in fillDescriptions).

See the presentation https://indico.cern.ch/event/664043/contributions/2713101/attachments/1519347/2372809/hcal_raddam_phase2_status_sept_6_2017.pdf and this plot that demonstrates improvement in jet energy response:
JER w/ M2 fix

This PR will be backported to 93X for inclusion in the HGCal TDR release (hopefully).

attn: @abdoulline @mariadalfonso @hatakeyamak

### CODE REVIEW

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
please test

**Comment by slava77** [comment]
assign upgrade
at least it will be easier to search for upgrade-related PRs

**Comment by kpedro88** [comment]
Minor changes in HCAL reconstruction for Phase2 workflows, as expected; no changes otherwise. Examples:

log10(chi2)
log10(energy)

**Comment by kpedro88** [comment]
+1

**Comment by kpedro88** [comment]
Further demonstration of effectiveness of the fix: restoration of expected JER
JER

**Comment by Martin-Grunewald** [comment]
+1

File: RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc [hunk 1]
```diff
@@ -573,10 +575,12 @@ void PulseShapeFitOOTPileupCorrection::phase1Apply(const HBHEChannelInfo& channe
     }
   }
 
-  double averagePedSig2GeV=0.25*(channelData.tsPedestalWidth(0)*channelData.tsPedestalWidth(0)*channelData.tsGain(0)*channelData.tsGain(0) +
-				 channelData.tsPedestalWidth(1)*channelData.tsPedestalWidth(1)*channelData.tsGain(1)*channelData.tsGain(1) +
-				 channelData.tsPedestalWidth(2)*channelData.tsPedestalWidth(2)*channelData.tsGain(2)*channelData.tsGain(2) +
-				 channelData.tsPedestalWidth(3)*channelData.tsPedestalWidth(3)*channelData.tsGain(3)*channelData.tsGain(3));
+  double sipmDarkCurrentWidth2 = 0.;
+  if(dcConstraint_) sipmDarkCurrentWidth2 = sipmDarkCurrentWidth*sipmDarkCurrentWidth;
```
Comment (slava77) [inline RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc]: IIRC, the dark current distribution is substantially non-Gaussian. So, the constraint done this way is perhaps not the best solution.
If so, I guess it's still OK for a trial for phase-2 studies.

Comment (kpedro88) [reply RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc]: yes, that will need to be a long-term development item for M2.

**Comment by slava77** [comment]
@kpedro88 
has there been much progress towards implementing the fillDescriptions for HCAL local reco?

**Comment by slava77** [comment]
@kpedro88 
chi2 in the plots in 2023 workflows in jenkins didn't change much.
Is this because the default inputs have no aging (and the effect of the dark current is small)?

**Comment by kpedro88** [comment]
@slava77 
1. You'll have to ask @igv4321 about fillDescriptions for HCAL reco code; well beyond the scope of this PR, and also I'm not even sure if it makes sense with an algorithm plugin system like the one we have.
2. Yes, the default inputs have no aging, so the effect of slightly reducing the constraint from the third term of M2 is small. The default dark current is tiny (less than an MeV). If you look at the presentation I linked in the PR description, you can see the effect of the fix with aging (slide 9).

**Comment by slava77** [comment]
On 9/11/17 11:16 AM, Kevin Pedro wrote:
> @slava77 <https://github.com/slava77>
> 
>  1. You'll have to ask @igv4321 <https://github.com/igv4321> about
>     fillDescriptions for HCAL reco code; well beyond the scope of this
>     PR, and also I'm not even sure if it makes sense with an algorithm
>     plugin system like the one we have.

OK. This wasn't for this PR, more of a check on the status of this.
There are solutions (check tracking code) where the fillDescriptions was
eventually made feasible.

>  2. Yes, the default inputs have no aging, so the effect of slightly
>     reducing the constraint from the third term of M2 is small. The
>     default dark current is tiny (less than an MeV). If you look at the
>     presentation I linked in the PR description, you can see the effect
>     of the fix with aging (slide 9).

I saw the slides had large changes in chi2, but not the jenkins diffs.
Lack of aging by default explains what's happening.
Thanks.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/20422	5051416
- changes are in line with the description: add a switch and use it to introduce a dark current "width" term in the M2 chi2. Done only for phase-2 .
- jenkins tests pass and comparisons with baseline show differences only in phase-2 wf starting from HBHE rechits which propagate downstream. 
- More details in support of the signature are in https://indico.cern.ch/event/664043/contributions/2713101/attachments/1519347/2372809/hcal_raddam_phase2_status_sept_6_2017.pdf provided with the PR

**Comment by davidlange6** [comment]
+1

**Pull request merged**